File: .Test-Demo/PhotoAssessmentUsageDemo/PhotoAssessmentUsageDemo/PhotoAssessmentHelper.m
//
//  PhotoAssessmentHelper.m
//  PhotoAssessmentUsageDemo
//
//  Created by 杨萧玉 on 2018/11/30.
//  Copyright © 2018 杨萧玉. All rights reserved.
//

#import "PhotoAssessmentHelper.h"
#import "PhotoAssessmentUtils.h"
#import "PhotoMPSProcessor.h"
#import "PhotoMLProcessor.h"
#import <Metal/Metal.h>

API_AVAILABLE(ios(11.0))
@interface PhotoAssessmentHelper ()

@property (nonatomic, strong) PhotoMPSProcessor *mpsProcessor;
@property (nonatomic, strong) PhotoMLProcessor *mlProcessor;
@property (nonatomic, strong) dispatch_queue_t processQueue;

@end

@implementation PhotoAssessmentHelper

- (instancetype)init
{
    self = [super init];
    if (self) {
        if (@available(iOS 11.0, *)) {
            _mpsProcessor = [PhotoMPSProcessor new];
            _mlProcessor = [PhotoMLProcessor new];
        } else {
            // Fallback on earlier versions
        }
        _processQueue = dispatch_queue_create("com.photoassessment.helper", DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

- (void)requestMLAssessmentScoreFor:(CGImageRef)image completionHandler:(void(^)(double))completionHandler
{
    return [self.mlProcessor processImage:image completionHandler:completionHandler];
}

- (void)requestMPSAssessmentScoreFor:(CGImageRef)image completionHandler:(void (^)(PhotoAssessmentResult *))completionHandler
{
    if (!completionHandler) {
        return;
    }
    CGImageRetain(image);
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
        
        NSUInteger width = CGImageGetWidth(image);
        NSUInteger height = CGImageGetHeight(image);
        int32_t *pixels = (int32_t *)calloc(height * width, sizeof(int32_t));
        CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
        CGContextRef context = CGBitmapContextCreate(pixels, width, height,
                                                     8, 4 * width, colorSpace,
                                                     kCGImageAlphaPremultipliedLast|kCGBitmapByteOrder32Little);
        CGContextDrawImage(context, CGRectMake(0, 0, width, height), image);
        
        
        PhotoAssessmentResult *totalResult = [PhotoAssessmentResult new];
        dispatch_group_t group = dispatch_group_create();
        CGFloat side = 32;
        dispatch_group_enter(group);
        [self.mlProcessor processImage:image completionHandler:^(double score) {
            totalResult.contentScore = score;
            dispatch_group_leave(group);
        }];
        
        CGImageRelease(image);
        dispatch_group_enter(group);
        [self.mpsProcessor downsampleWithImagePixels:pixels width:width height:height scaleDimension:side completionHandler:^(NSArray<NSNumber *> * _Nullable result) {
            if (result) {
                NSDictionary<NSNumber *, NSNumber *> *fingerprint = [PhotoAssessmentUtils fingerprintForImagePixels:result width:side height:side];
                HSBColor *hsb = [PhotoAssessmentUtils meanHSBForImagePixels:result width:side height:side];
                dispatch_async(self.processQueue, ^{
                    totalResult.fingerprint = fingerprint;
                    totalResult.hsb = hsb;
                    dispatch_group_leave(group);
                });
            }
            else {
                dispatch_group_leave(group);
            }
        }];
        dispatch_group_enter(group);
        [self.mpsProcessor edgeDetectWithImagePixels:pixels width:width height:height completionHandler:^(int8_t mean, int8_t variance) {
            dispatch_async(self.processQueue, ^{
                totalResult.edgeDetectMean = mean;
                totalResult.edgeDetectVariance = variance;
                dispatch_group_leave(group);
            });
        }];
        
        CGColorSpaceRelease(colorSpace);
        CGContextRelease(context);
        free(pixels);
        dispatch_group_notify(group, self.processQueue, ^{
            completionHandler(totalResult);
        });
    });
}

@end

@implementation PhotoAssessmentResult

- (NSString *)description
{
    return [NSString stringWithFormat:@"EdgeDetect mean:%d, variance:%d\nHSB: h(%.3f), s(%.3f), b(%.3f)\nContentScore:%f", self.edgeDetectMean, self.edgeDetectVariance, self.hsb.hue, self.hsb.saturation, self.hsb.brightness, self.contentScore];
}

+ (BOOL)supportsSecureCoding
{
    return YES;
}

- (void)encodeWithCoder:(nonnull NSCoder *)aCoder {
    [aCoder encodeInteger:self.edgeDetectMean forKey:@"edgeDetectMean"];
    [aCoder encodeInteger:self.edgeDetectVariance forKey:@"edgeDetectVariance"];
    [aCoder encodeObject:self.hsb forKey:@"hsb"];
    [aCoder encodeObject:self.fingerprint forKey:@"fingerprint"];
    [aCoder encodeDouble:self.contentScore forKey:@"contentScore"];
}

- (nullable instancetype)initWithCoder:(nonnull NSCoder *)aDecoder {
    if (self = [self init]) {
        self.edgeDetectMean = [aDecoder decodeIntegerForKey:@"edgeDetectMean"];
        self.edgeDetectVariance = [aDecoder decodeIntegerForKey:@"edgeDetectVariance"];
        self.hsb = [aDecoder decodeObjectForKey:@"hsb"];
        self.fingerprint = [aDecoder decodeObjectForKey:@"fingerprint"];
        self.contentScore = [aDecoder decodeDoubleForKey:@"contentScore"];
    }
    return self;
}

- (double)totalScore
{
//    return self.contentScore;
    return self.hsb.saturation * (self.edgeDetectMean + self.edgeDetectVariance);
}

- (BOOL)betterThan:(PhotoAssessmentResult *)anotherResult
{
    double scoreA = [self totalScore];
    double scoreB = [anotherResult totalScore];
    return scoreA > scoreB;
}

@end

File: PhotoAssessment-iOSSample/PhotoAssessment/PhotoAssessmentHelper.swift
//
//  PhotoAssessmentHelper.swift
//  PhotoAssessment
//
//  Created by 杨萧玉 on 2018/11/19.
//  Copyright © 2018 杨萧玉. All rights reserved.
//

import UIKit

open class PhotoAssessmentResult: NSObject, NSCoding {
    
    @objc public var edgeDetectMean: Int8
    @objc public var edgeDetectVariance: Int8
    @objc public var hsb: HSBColor?
    @objc public var fingerprint: [UInt32: UInt]?
    @objc public var contentScore: Double
    
    override init() {
        edgeDetectMean = 0
        edgeDetectVariance = 0
        contentScore = 0
    }
    
    open override var description: String {
        var text = ""
        text += "edgeDetect mean: \(String(describing: edgeDetectMean)) variance: \(String(describing: edgeDetectVariance))"
        if let hsb = hsb {
            text += String(format: "\nhsb: h(%.3f), s(%.3f), b(%.3f)", hsb.hue, hsb.saturation, hsb.brightness)
        }
        text += String(format: "\ncontentScore: %.3f", contentScore)
        return text
    }
    
    public func encode(with aCoder: NSCoder) {
        aCoder.encode(self.edgeDetectMean, forKey: "edgeDetectMean")
        aCoder.encode(self.edgeDetectVariance, forKey: "edgeDetectVariance")
        aCoder.encode(self.hsb, forKey: "hsb")
        aCoder.encode(self.fingerprint, forKey: "fingerprint")
        aCoder.encode(self.contentScore, forKey: "contentScore")
    }
    
    public required init?(coder aDecoder: NSCoder) {
        self.edgeDetectMean = aDecoder.decodeObject(forKey: "edgeDetectMean") as? Int8 ?? 0
        self.edgeDetectVariance = aDecoder.decodeObject(forKey: "edgeDetectVariance") as? Int8 ?? 0
        self.hsb = aDecoder.decodeObject(forKey: "hsb") as? HSBColor
        self.fingerprint = aDecoder.decodeObject(forKey: "fingerprint") as? [UInt32: UInt]
        self.contentScore = aDecoder.decodeObject(forKey: "contentScore") as? Double ?? 0
    }
}

@available(iOS 11.0, macOS 10.13, tvOS 11.0, *)
open class PhotoAssessmentHelper: NSObject {
    
    private let mpsProcessor = PhotoMPSProcessor()
    private let mlProcessor = PhotoMLProcessor()
    private let processQueue = DispatchQueue(label: "com.photoassessment.helper")
    
    @objc public func requestMLAssessmentScore(for image: CGImage, completionHandler: @escaping (Double) -> Void) {
        let start = Date()
        mlProcessor.process(image: image, completionHandler: { (score) in
            print("ml process duration:\(Date().timeIntervalSince(start))")
            completionHandler(score)
        })
    }
    
    @objc public func requestMPSAssessmentScore(for image: CGImage, completionHandler: @escaping (PhotoAssessmentResult) -> Void) {
        var start = Date()
        let imagePixels = image.rgbPixels()
        print("rgb pixels duration:\(Date().timeIntervalSince(start))")
        let totalResult = PhotoAssessmentResult()
        DispatchQueue.main.async {
            start = Date()
            let side = 50
            let group = DispatchGroup()
            group.enter()
            self.mpsProcessor.downsample(imagePixels: imagePixels, width: image.width, height: image.height, scaleDimension: side, { (result) in
                if let pixels = result {
                    let fingerprint = Utils.fingerprintFor(imagePixels: pixels, width: side, height: side)
                    print("finger print duration:\(Date().timeIntervalSince(start))")
                    
                    start = Date()
                    let hsb = Utils.meanHSBFor(imagePixels: pixels, width: side, height: side)
                    print("hsb duration:\(Date().timeIntervalSince(start))")
                    
                    self.processQueue.async {
                        totalResult.fingerprint = fingerprint
                        totalResult.hsb = hsb
                        group.leave()
                    }
                }
                else {
                    group.leave()
                }
            })
            group.enter()
            start = Date()
            self.mpsProcessor.edgeDetect(imagePixels: imagePixels, width: image.width, height: image.height, { (mean, variance) in
                print("fuzzy degree duration:\(Date().timeIntervalSince(start))")
                self.processQueue.async {
                    totalResult.edgeDetectMean = mean
                    totalResult.edgeDetectVariance = variance
                    group.leave()
                }
            })
            
            group.notify(queue: self.processQueue) {
                completionHandler(totalResult)
            }
        }
    }
}
